# Análisis del dataset MMM semanal (India)

Este notebook resume el comportamiento observado en `synthetic_mmm_weekly_india.csv` y traza los pasos para construir un modelo de Marketing Mix con esas variables. La idea es tener un punto de partida rápido para entender la granularidad, las palancas disponibles y la rigidez del KPI.


## Analisis exploratorio

In [19]:
"""
Load of the dataframe
"""
import pandas as pd
from pathlib import Path

pd.set_option("display.max_columns", None)
pd.set_option("display.width", None)

df_main.describe()


path = Path('/Users/juanmedina1810/Desktop/marketing_insights_project/data/raw/weekly_india/synthetic_mmm_weekly_india.csv')

df_main = pd.read_csv(path, parse_dates=['Week'])
df_main = df_main.sort_values('Week')

print(f'Shape: {df_main.shape}')
display(df_main.head())


Shape: (11232, 28)


Week        Geo   Brand          SKU  Sales_Units   Sales_Value  \
0  2022-07-04    CENTRAL  BrandA  BrandA_SKU1   106.327944  10551.108260   
51 2022-07-04  NORTHEAST  BrandC  BrandC_SKU1    86.890420   5022.978068   
50 2022-07-04  NORTHEAST  BrandB  BrandB_SKU3    26.414406   2159.158135   
49 2022-07-04  NORTHEAST  BrandB  BrandB_SKU2   128.322961  10008.159352   
48 2022-07-04  NORTHEAST  BrandB  BrandB_SKU1    63.865638   4606.745823   

          MRP  Net_Price  Feature_Flag  Display_Flag  TPR_Flag   Trade_Spend  \
0   99.231753  99.231753             0             0         0  12108.945933   
51  57.808192  57.808192             0             0         0  13616.569652   
50  81.741689  81.741689             0             0         0  13616.569652   
49  77.991961  77.991961             0             0         0  13616.569652   
48  72.131838  72.131838             0             0         0  13616.569652   

    TV_Impressions  YouTube_Impressions  Facebook_Impressions  \
0     1.097593e+06        526732.456845         476584.682221   
51    2.750571e+05        120199.021228          97337.569229   
50    5.910005e+05        214922.930575         177730.462181   
49    3.534878e+05        129222.735250         122089.485936   
48    6.219886e+05        257125.565333         206715.038120   

    Instagram_Impressions  Print_Readership  Radio_Listenership  \
0           153004.959591      78906.938696        1.050008e+06   
51           33773.561168      22089.165928        2.577155e+05   
50           60757.632333      41080.549314        4.455845e+05   
49           38053.495644      27028.458770        2.705343e+05   
48           72096.616879      45374.513866        4.866983e+05   

    FB_Banner_Content_Score  IG_Banner_Content_Score  Weighted_Distribution  \
0                 74.714021                77.999979               0.326712   
51                73.725344                76.223658               0.434818   
50                73.725344                76.223658               0.434818   
49                73.725344                76.223658               0.434818   
48                73.725344                76.223658               0.434818   

    Numeric_Distribution        TDP       NOS         CPI  GDP_Growth  \
0               0.406848  36.923185  3.071753  120.954667     5.24444   
51              0.430918  36.029924  3.277555  115.377600     5.07394   
50              0.430918  36.029924  3.277555  115.377600     5.07394   
49              0.430918  36.029924  3.277555  115.377600     5.07394   
48              0.430918  36.029924  3.277555  115.377600     5.07394   

    Festival_Index  Rainfall_Index  
0         1.180806        5.378450  
51        0.855645        4.729768  
50        0.855645        4.729768  
49        0.855645        4.729768  
48        0.855645        4.729768

In [22]:
"""
List of unique columns with description
"""
df.describe(include="all").T

count unique          top  freq  \
Week                       11232    NaN          NaN   NaN   
Geo                        11232      8      CENTRAL  1404   
Brand                      11232      3       BrandA  3744   
SKU                        11232      9  BrandA_SKU1  1248   
Sales_Units              11232.0    NaN          NaN   NaN   
Sales_Value              11232.0    NaN          NaN   NaN   
MRP                      11232.0    NaN          NaN   NaN   
Net_Price                11232.0    NaN          NaN   NaN   
Feature_Flag             11232.0    NaN          NaN   NaN   
Display_Flag             11232.0    NaN          NaN   NaN   
TPR_Flag                 11232.0    NaN          NaN   NaN   
Trade_Spend              11232.0    NaN          NaN   NaN   
TV_Impressions           11232.0    NaN          NaN   NaN   
YouTube_Impressions      11232.0    NaN          NaN   NaN   
Facebook_Impressions     11232.0    NaN          NaN   NaN   
Instagram_Impressions    11232.0    NaN          NaN   NaN   
Print_Readership         11232.0    NaN          NaN   NaN   
Radio_Listenership       11232.0    NaN          NaN   NaN   
FB_Banner_Content_Score  11232.0    NaN          NaN   NaN   
IG_Banner_Content_Score  11232.0    NaN          NaN   NaN   
Weighted_Distribution    11232.0    NaN          NaN   NaN   
Numeric_Distribution     11232.0    NaN          NaN   NaN   
TDP                      11232.0    NaN          NaN   NaN   
NOS                      11232.0    NaN          NaN   NaN   
CPI                      11232.0    NaN          NaN   NaN   
GDP_Growth               11232.0    NaN          NaN   NaN   
Festival_Index           11232.0    NaN          NaN   NaN   
Rainfall_Index           11232.0    NaN          NaN   NaN   

                                        mean                  min  \
Week                     2023-12-28 12:00:00  2022-07-04 00:00:00   
Geo                                      NaN                  NaN   
Brand                                    NaN                  NaN   
SKU                                      NaN                  NaN   
Sales_Units                       422.432277            25.535563   
Sales_Value                     33099.052755          2054.711829   
MRP                                79.911714            52.759457   
Net_Price                            78.3536            46.696772   
Feature_Flag                        0.185096                  0.0   
Display_Flag                        0.183494                  0.0   
TPR_Flag                            0.157853                  0.0   
Trade_Spend                     53976.998415          6140.424107   
TV_Impressions                1043587.224538         102622.70648   
YouTube_Impressions            534205.428103         46338.661619   
Facebook_Impressions           320951.858958         14170.185475   
Instagram_Impressions          138936.271516         12882.596915   
Print_Readership                85324.757262         10148.344288   
Radio_Listenership              317198.63882         23387.680107   
FB_Banner_Content_Score            70.340914            46.951991   
IG_Banner_Content_Score            65.260831            29.567246   
Weighted_Distribution               0.621472             0.304993   
Numeric_Distribution                0.655593             0.285294   
TDP                                67.510443            29.143297   
NOS                                 4.419803             2.652576   
CPI                                 130.8247           114.561767   
GDP_Growth                           4.95127             3.653294   
Festival_Index                      1.538062             0.855645   
Rainfall_Index                      2.049459             0.323376   

                                         25%                  50%  \
Week                     2023-04-01 06:00:00  2023-12-28 12:00:00   
Geo                                      NaN                  NaN   
Brand              

### Analisis exploratorio por variable

In [ ]:
"""
Unique values in categorical columns
"""

for col in ["Geo", "Brand", "SKU"]:
    values = sorted(df_main[col].dropna().unique())
    print(f"\n{col} ({len(values)} categorías):")
    print(values)


Geo (8 categorías):
['CENTRAL', 'EAST', 'METRO_DELHI', 'METRO_MUMBAI', 'NORTH', 'NORTHEAST', 'SOUTH', 'WEST']

Brand (3 categorías):
['BrandA', 'BrandB', 'BrandC']

SKU (9 categorías):
['BrandA_SKU1', 'BrandA_SKU2', 'BrandA_SKU3', 'BrandB_SKU1', 'BrandB_SKU2', 'BrandB_SKU3', 'BrandC_SKU1', 'BrandC_SKU2', 'BrandC_SKU3']


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import ipywidgets as widgets
from IPython.display import display, clear_output

# --- Config ---
df = df_main.copy()

time_col_default = "Week"

num_cols = df.select_dtypes(include=np.number).columns.tolist()

# Paleta simple y consistente
COLOR_PALETTE = {
    "Azul": "tab:blue",
    "Naranja": "tab:orange",
    "Verde": "tab:green",
    "Rojo": "tab:red",
    "Violeta": "tab:purple",
    "Marrón": "tab:brown"
}

def _safe_sort_key(series: pd.Series):
    try:
        return pd.to_numeric(series)
    except Exception:
        return series.astype(str)

# --- Widgets ---
w_time = widgets.Dropdown(
    options=df.columns.tolist(),
    value=time_col_default if time_col_default in df.columns else df.columns[0],
    description="Time:"
)

w_y1 = widgets.Dropdown(
    options=num_cols,
    description="Y1:"
)

w_y2 = widgets.Dropdown(
    options=["(none)"] + num_cols,
    value="(none)",
    description="Y2:"
)

w_agg = widgets.Dropdown(
    options=["mean", "sum", "median"],
    value="mean",
    description="Agg:"
)

w_color_y1 = widgets.Dropdown(
    options=COLOR_PALETTE,
    value="tab:blue",
    description="Color Y1:"
)

w_color_y2 = widgets.Dropdown(
    options=COLOR_PALETTE,
    value="tab:orange",
    description="Color Y2:"
)

w_secondary = widgets.Checkbox(
    value=True,
    description="Y2 eje secundario"
)

w_grid = widgets.Checkbox(
    value=True,
    description="Grid"
)

out = widgets.Output()

def plot_timeseries(_=None):
    with out:
        clear_output(wait=True)

        time_col = w_time.value
        y1 = w_y1.value
        y2 = None if w_y2.value == "(none)" else w_y2.value
        agg = w_agg.value

        if y1 is None:
            print("No hay columnas numéricas para graficar.")
            return

        cols = [time_col, y1] + ([y2] if y2 else [])
        d = df[cols].dropna(subset=[time_col]).copy()

        g = d.groupby(time_col, dropna=False).agg(agg).reset_index()
        g = g.sort_values(by=time_col, key=_safe_sort_key)

        fig, ax1 = plt.subplots(figsize=(12, 5))

        ax1.plot(
            g[time_col],
            g[y1],
            marker="o",
            color=w_color_y1.value,
            label=y1
        )
        ax1.set_ylabel(y1)
        ax1.set_xlabel(time_col)

        if w_grid.value:
            ax1.grid(True)

        if y2:
            if w_secondary.value:
                ax2 = ax1.twinx()
                ax2.plot(
                    g[time_col],
                    g[y2],
                    marker="o",
                    linestyle="--",
                    color=w_color_y2.value,
                    label=y2
                )
                ax2.set_ylabel(y2)
            else:
                ax1.plot(
                    g[time_col],
                    g[y2],
                    marker="o",
                    linestyle="--",
                    color=w_color_y2.value,
                    label=y2
                )
                ax1.set_ylabel(f"{y1} / {y2}")

        if g[time_col].dtype == "object":
            plt.xticks(rotation=45, ha="right")

        plt.title(f"{agg} por {time_col}: {y1}" + (f" vs {y2}" if y2 else ""))
        plt.tight_layout()
        plt.show()

# Eventos
for w in [
    w_time, w_y1, w_y2, w_agg,
    w_color_y1, w_color_y2,
    w_secondary, w_grid
]:
    w.observe(plot_timeseries, names="value")

# UI
ui = widgets.VBox([
    widgets.HBox([w_time, w_agg]),
    widgets.HBox([w_y1, w_color_y1]),
    widgets.HBox([w_y2, w_color_y2]),
    widgets.HBox([w_secondary, w_grid]),
    out
])

display(ui)
plot_timeseries()


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import ipywidgets as widgets
from IPython.display import display, clear_output

df = df_main.copy()

time_col_default = "Week"

# columnas numéricas
num_cols = df.select_dtypes(include=np.number).columns.tolist()

# paleta simple
COLOR_PALETTE = {
    "Azul": "tab:blue",
    "Naranja": "tab:orange",
    "Verde": "tab:green",
    "Rojo": "tab:red",
}

def _safe_sort_key(series: pd.Series):
    try:
        return pd.to_numeric(series)
    except Exception:
        return series.astype(str)

def _sorted_unique(col):
    return sorted(df[col].dropna().unique().tolist())

# --- Widgets base ---
w_time = widgets.Dropdown(
    options=df.columns.tolist(),
    value=time_col_default if time_col_default in df.columns else df.columns[0],
    description="Time:"
)

w_agg = widgets.Dropdown(
    options=["mean", "sum", "median"],
    value="mean",
    description="Agg:"
)

w_y1 = widgets.Dropdown(
    options=num_cols,
    value=num_cols[0] if num_cols else None,
    description="Y1:"
)

w_y2 = widgets.Dropdown(
    options=["(none)"] + num_cols,
    value="(none)",
    description="Y2:"
)

w_color_y1 = widgets.Dropdown(
    options=COLOR_PALETTE,
    value="tab:blue",
    description="Color Y1:"
)

w_color_y2 = widgets.Dropdown(
    options=COLOR_PALETTE,
    value="tab:orange",
    description="Color Y2:"
)

w_secondary = widgets.Checkbox(value=True, description="Y2 eje secundario")
w_grid = widgets.Checkbox(value=True, description="Grid")

# --- Filtros categóricos (multi-select) ---
w_geo = widgets.SelectMultiple(
    options=_sorted_unique("Geo") if "Geo" in df.columns else [],
    value=tuple(_sorted_unique("Geo")) if "Geo" in df.columns else tuple(),
    description="Geo",
    rows=8
)

w_brand = widgets.SelectMultiple(
    options=_sorted_unique("Brand") if "Brand" in df.columns else [],
    value=tuple(_sorted_unique("Brand")) if "Brand" in df.columns else tuple(),
    description="Brand",
    rows=5
)

w_sku = widgets.SelectMultiple(
    options=_sorted_unique("SKU") if "SKU" in df.columns else [],
    value=tuple(_sorted_unique("SKU")) if "SKU" in df.columns else tuple(),
    description="SKU",
    rows=10
)

# Botones select all / clear
btn_geo_all = widgets.Button(description="All Geo", layout=widgets.Layout(width="90px"))
btn_geo_clear = widgets.Button(description="Clear", layout=widgets.Layout(width="90px"))

btn_brand_all = widgets.Button(description="All Brand", layout=widgets.Layout(width="90px"))
btn_brand_clear = widgets.Button(description="Clear", layout=widgets.Layout(width="90px"))

btn_sku_all = widgets.Button(description="All SKU", layout=widgets.Layout(width="90px"))
btn_sku_clear = widgets.Button(description="Clear", layout=widgets.Layout(width="90px"))

def _set_all(select_widget):
    # en SelectMultiple, value tiene que ser tuple
    select_widget.value = tuple(select_widget.options)

def _clear(select_widget):
    select_widget.value = tuple()

btn_geo_all.on_click(lambda _: _set_all(w_geo))
btn_geo_clear.on_click(lambda _: _clear(w_geo))

btn_brand_all.on_click(lambda _: _set_all(w_brand))
btn_brand_clear.on_click(lambda _: _clear(w_brand))

btn_sku_all.on_click(lambda _: _set_all(w_sku))
btn_sku_clear.on_click(lambda _: _clear(w_sku))

out = widgets.Output()

def plot_timeseries(_=None):
    with out:
        clear_output(wait=True)

        time_col = w_time.value
        y1 = w_y1.value
        y2 = None if w_y2.value == "(none)" else w_y2.value
        agg = w_agg.value

        if y1 is None:
            print("No hay columnas numéricas para graficar.")
            return

        # --- aplicar filtros ---
        d = df.copy()

        if "Geo" in d.columns:
            if len(w_geo.value) == 0:
                print("Geo está vacío. Seleccioná al menos 1 Geo o tocá 'All Geo'.")
                return
            d = d[d["Geo"].isin(list(w_geo.value))]

        if "Brand" in d.columns:
            if len(w_brand.value) == 0:
                print("Brand está vacío. Seleccioná al menos 1 Brand o tocá 'All Brand'.")
                return
            d = d[d["Brand"].isin(list(w_brand.value))]

        if "SKU" in d.columns:
            if len(w_sku.value) == 0:
                print("SKU está vacío. Seleccioná al menos 1 SKU o tocá 'All SKU'.")
                return
            d = d[d["SKU"].isin(list(w_sku.value))]

        cols = [time_col, y1] + ([y2] if y2 else [])
        d = d[cols].dropna(subset=[time_col]).copy()

        # agrupar por semana y agregar
        g = d.groupby(time_col, dropna=False).agg(agg).reset_index()
        g = g.sort_values(by=time_col, key=_safe_sort_key)

        # plot
        fig, ax1 = plt.subplots(figsize=(12, 5))

        ax1.plot(
            g[time_col], g[y1],
            marker="o",
            color=w_color_y1.value,
            label=y1
        )
        ax1.set_ylabel(y1)
        ax1.set_xlabel(time_col)

        if w_grid.value:
            ax1.grid(True)

        if y2:
            if w_secondary.value:
                ax2 = ax1.twinx()
                ax2.plot(
                    g[time_col], g[y2],
                    marker="o",
                    linestyle="--",
                    color=w_color_y2.value,
                    label=y2
                )
                ax2.set_ylabel(y2)
            else:
                ax1.plot(
                    g[time_col], g[y2],
                    marker="o",
                    linestyle="--",
                    color=w_color_y2.value,
                    label=y2
                )
                ax1.set_ylabel(f"{y1} / {y2}")

        if g[time_col].dtype == "object":
            plt.xticks(rotation=45, ha="right")

        # título con filtros (resumen corto)
        geo_txt = f"Geo={len(w_geo.value)}" if "Geo" in df.columns else ""
        brand_txt = f"Brand={len(w_brand.value)}" if "Brand" in df.columns else ""
        sku_txt = f"SKU={len(w_sku.value)}" if "SKU" in df.columns else ""
        filt_txt = ", ".join([t for t in [geo_txt, brand_txt, sku_txt] if t])

        plt.title(f"{agg} por {time_col}: {y1}" + (f" vs {y2}" if y2 else "") + (f" | {filt_txt}" if filt_txt else ""))
        plt.tight_layout()
        plt.show()

# observers
for w in [
    w_time, w_agg, w_y1, w_y2,
    w_color_y1, w_color_y2,
    w_secondary, w_grid,
    w_geo, w_brand, w_sku
]:
    w.observe(plot_timeseries, names="value")

# UI layout
filters_ui = widgets.HBox([
    widgets.VBox([w_geo, widgets.HBox([btn_geo_all, btn_geo_clear])]),
    widgets.VBox([w_brand, widgets.HBox([btn_brand_all, btn_brand_clear])]),
    widgets.VBox([w_sku, widgets.HBox([btn_sku_all, btn_sku_clear])]),
])

controls_ui = widgets.VBox([
    widgets.HBox([w_time, w_agg]),
    widgets.HBox([w_y1, w_color_y1]),
    widgets.HBox([w_y2, w_color_y2]),
    widgets.HBox([w_secondary, w_grid]),
])

ui = widgets.VBox([filters_ui, controls_ui, out])

display(ui)
plot_timeseries()



In [33]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import ipywidgets as widgets
from IPython.display import display, clear_output

# =========================
# Config / Data
# =========================
df = df_main.copy()

time_col_default = "Week"
geo_col = "Geo"
brand_col = "Brand"
sku_col = "SKU"

# Asegurar datetime en Week (si ya es datetime, no rompe)
if time_col_default in df.columns:
    df[time_col_default] = pd.to_datetime(df[time_col_default], errors="coerce")

# Columnas numéricas
num_cols = df.select_dtypes(include=np.number).columns.tolist()

# Paleta simple (4 colores, como pediste)
COLOR_PALETTE = {
    "Azul": "tab:blue",
    "Naranja": "tab:orange",
    "Verde": "tab:green",
    "Rojo": "tab:red",
}

def _sorted_unique(col):
    if col not in df.columns:
        return []
    return sorted(df[col].dropna().unique().tolist())

# =========================
# Widgets base
# =========================
w_time = widgets.Dropdown(
    options=df.columns.tolist(),
    value=time_col_default if time_col_default in df.columns else df.columns[0],
    description="Time:"
)

w_agg = widgets.Dropdown(
    options=["mean", "sum", "median"],
    value="mean",
    description="Agg:"
)

w_y1 = widgets.Dropdown(
    options=num_cols,
    value=num_cols[0] if num_cols else None,
    description="Y1:"
)

w_y2 = widgets.Dropdown(
    options=["(none)"] + num_cols,
    value="(none)",
    description="Y2:"
)

w_color_y1 = widgets.Dropdown(
    options=COLOR_PALETTE,
    value="tab:blue",
    description="Color Y1:"
)

w_color_y2 = widgets.Dropdown(
    options=COLOR_PALETTE,
    value="tab:orange",
    description="Color Y2:"
)

w_secondary = widgets.Checkbox(value=True, description="Y2 eje secundario")
w_grid = widgets.Checkbox(value=True, description="Grid")
w_year_lines = widgets.Checkbox(value=True, description="Líneas año")

# =========================
# Filtros categóricos (multi)
# =========================
w_geo = widgets.SelectMultiple(
    options=_sorted_unique(geo_col),
    value=tuple(_sorted_unique(geo_col)),
    description="Geo",
    rows=8
)

w_brand = widgets.SelectMultiple(
    options=_sorted_unique(brand_col),
    value=tuple(_sorted_unique(brand_col)),
    description="Brand",
    rows=5
)

w_sku = widgets.SelectMultiple(
    options=_sorted_unique(sku_col),
    value=tuple(_sorted_unique(sku_col)),
    description="SKU",
    rows=10
)

def _set_all(select_widget):
    select_widget.value = tuple(select_widget.options)

def _clear(select_widget):
    select_widget.value = tuple()

btn_geo_all = widgets.Button(description="All Geo", layout=widgets.Layout(width="90px"))
btn_geo_clear = widgets.Button(description="Clear", layout=widgets.Layout(width="90px"))
btn_geo_all.on_click(lambda _: _set_all(w_geo))
btn_geo_clear.on_click(lambda _: _clear(w_geo))

btn_brand_all = widgets.Button(description="All Brand", layout=widgets.Layout(width="90px"))
btn_brand_clear = widgets.Button(description="Clear", layout=widgets.Layout(width="90px"))
btn_brand_all.on_click(lambda _: _set_all(w_brand))
btn_brand_clear.on_click(lambda _: _clear(w_brand))

btn_sku_all = widgets.Button(description="All SKU", layout=widgets.Layout(width="90px"))
btn_sku_clear = widgets.Button(description="Clear", layout=widgets.Layout(width="90px"))
btn_sku_all.on_click(lambda _: _set_all(w_sku))
btn_sku_clear.on_click(lambda _: _clear(w_sku))

# =========================
# Filtros temporales (Year / Month)
# =========================
time_col = w_time.value
if time_col in df.columns and pd.api.types.is_datetime64_any_dtype(df[time_col]):
    df["_Year"] = df[time_col].dt.year
    df["_Month"] = df[time_col].dt.month
    years = sorted(df["_Year"].dropna().unique().astype(int).tolist())
else:
    df["_Year"] = np.nan
    df["_Month"] = np.nan
    years = []

months = list(range(1, 13))

w_year = widgets.SelectMultiple(
    options=years,
    value=tuple(years),
    description="Year",
    rows=min(6, max(1, len(years)))
)

w_month = widgets.SelectMultiple(
    options=months,
    value=tuple(months),
    description="Month",
    rows=6
)

btn_year_all = widgets.Button(description="All Year", layout=widgets.Layout(width="90px"))
btn_year_clear = widgets.Button(description="Clear", layout=widgets.Layout(width="90px"))
btn_year_all.on_click(lambda _: _set_all(w_year))
btn_year_clear.on_click(lambda _: _clear(w_year))

btn_month_all = widgets.Button(description="All Month", layout=widgets.Layout(width="90px"))
btn_month_clear = widgets.Button(description="Clear", layout=widgets.Layout(width="90px"))
btn_month_all.on_click(lambda _: _set_all(w_month))
btn_month_clear.on_click(lambda _: _clear(w_month))

# =========================
# Plot
# =========================
out = widgets.Output()

def plot_timeseries(_=None):
    with out:
        clear_output(wait=True)

        time_col = w_time.value
        y1 = w_y1.value
        y2 = None if w_y2.value == "(none)" else w_y2.value
        agg = w_agg.value

        if y1 is None:
            print("No hay columnas numéricas para graficar.")
            return

        # Copia para filtrar
        d = df.copy()

        # Validar que time_col sea datetime (necesario para Year/Month y líneas)
        if time_col not in d.columns:
            print(f"No existe la columna de tiempo: {time_col}")
            return

        d[time_col] = pd.to_datetime(d[time_col], errors="coerce")
        d = d.dropna(subset=[time_col])

        # Recalcular year/month por si el user cambió la columna time
        d["_Year"] = d[time_col].dt.year
        d["_Month"] = d[time_col].dt.month

        # ---- filtros categóricos
        if geo_col in d.columns:
            if len(w_geo.value) == 0:
                print("Geo vacío. Seleccioná al menos 1 o tocá 'All Geo'.")
                return
            d = d[d[geo_col].isin(list(w_geo.value))]

        if brand_col in d.columns:
            if len(w_brand.value) == 0:
                print("Brand vacío. Seleccioná al menos 1 o tocá 'All Brand'.")
                return
            d = d[d[brand_col].isin(list(w_brand.value))]

        if sku_col in d.columns:
            if len(w_sku.value) == 0:
                print("SKU vacío. Seleccioná al menos 1 o tocá 'All SKU'.")
                return
            d = d[d[sku_col].isin(list(w_sku.value))]

        # ---- filtros temporales
        if len(w_year.value) == 0:
            print("Year vacío. Seleccioná al menos 1 o tocá 'All Year'.")
            return
        if len(w_month.value) == 0:
            print("Month vacío. Seleccioná al menos 1 o tocá 'All Month'.")
            return

        d = d[d["_Year"].isin(list(w_year.value)) & d["_Month"].isin(list(w_month.value))]

        # Si queda vacío, avisar
        if d.empty:
            print("No hay datos con los filtros actuales (Geo/Brand/SKU/Year/Month).")
            return

        cols = [time_col, y1] + ([y2] if y2 else [])
        d = d[cols].copy()

        # groupby por time_col (Week) -> agregación
        g = d.groupby(time_col, dropna=False).agg(agg).reset_index()
        g = g.sort_values(time_col)

        # plot
        fig, ax1 = plt.subplots(figsize=(12, 5))

        ax1.plot(
            g[time_col], g[y1],
            marker="o",
            color=w_color_y1.value,
            label=y1
        )
        ax1.set_xlabel(time_col)
        ax1.set_ylabel(y1)

        if w_grid.value:
            ax1.grid(True)

        if y2:
            if w_secondary.value:
                ax2 = ax1.twinx()
                ax2.plot(
                    g[time_col], g[y2],
                    marker="o",
                    linestyle="--",
                    color=w_color_y2.value,
                    label=y2
                )
                ax2.set_ylabel(y2)
            else:
                ax1.plot(
                    g[time_col], g[y2],
                    marker="o",
                    linestyle="--",
                    color=w_color_y2.value,
                    label=y2
                )
                ax1.set_ylabel(f"{y1} / {y2}")

        # líneas verticales para separar años (inicio de año)
        if w_year_lines.value:
            years_in_plot = sorted(pd.to_datetime(g[time_col]).dt.year.unique().tolist())
            for yr in years_in_plot[1:]:
                ax1.axvline(
                    pd.Timestamp(year=int(yr), month=1, day=1),
                    color="black",
                    linestyle=":",
                    linewidth=1,
                    alpha=0.9
                )

        plt.xticks(rotation=45, ha="right")

        # Título con resumen de filtros
        filt_txt = []
        if geo_col in df.columns:   filt_txt.append(f"Geo={len(w_geo.value)}")
        if brand_col in df.columns: filt_txt.append(f"Brand={len(w_brand.value)}")
        if sku_col in df.columns:   filt_txt.append(f"SKU={len(w_sku.value)}")
        filt_txt.append(f"Years={len(w_year.value)}")
        filt_txt.append(f"Months={len(w_month.value)}")

        plt.title(f"{agg} por {time_col}: {y1}" + (f" vs {y2}" if y2 else "") + " | " + ", ".join(filt_txt))
        plt.tight_layout()
        plt.show()

# observers
for w in [
    w_time, w_agg, w_y1, w_y2,
    w_color_y1, w_color_y2,
    w_secondary, w_grid, w_year_lines,
    w_geo, w_brand, w_sku,
    w_year, w_month
]:
    w.observe(plot_timeseries, names="value")

# =========================
# Layout UI
# =========================
filters_ui = widgets.HBox([
    widgets.VBox([w_geo, widgets.HBox([btn_geo_all, btn_geo_clear])]),
    widgets.VBox([w_brand, widgets.HBox([btn_brand_all, btn_brand_clear])]),
    widgets.VBox([w_sku, widgets.HBox([btn_sku_all, btn_sku_clear])]),
])

time_ui = widgets.HBox([
    widgets.VBox([w_year, widgets.HBox([btn_year_all, btn_year_clear])]),
    widgets.VBox([w_month, widgets.HBox([btn_month_all, btn_month_clear])]),
])

controls_ui = widgets.VBox([
    widgets.HBox([w_time, w_agg]),
    time_ui,
    widgets.HBox([w_y1, w_color_y1]),
    widgets.HBox([w_y2, w_color_y2]),
    widgets.HBox([w_secondary, w_grid, w_year_lines]),
])

ui = widgets.VBox([filters_ui, controls_ui, out])

display(ui)
plot_timeseries()


### Resultados de analisi exploratorio

In [10]:
channels = ['TV_Impressions', 'YouTube_Impressions', 'Facebook_Impressions', 'Instagram_Impressions']
print('Correlaciones con Sales_Value para los canales mas visibles:')
print(df[channels + ['Sales_Value']].corr()['Sales_Value'].drop('Sales_Value').sort_values(ascending=False))


Correlaciones con Sales_Value para los canales mas visibles:
Instagram_Impressions    0.476169
TV_Impressions           0.475807
YouTube_Impressions      0.393494
Facebook_Impressions     0.389887
Name: Sales_Value, dtype: float64


## Observaciones claves
- El dataset es semanal por `Week`, `Geo`, `Brand` y `SKU`, con 11k registros sinteticos.
- Las ventas (`Sales_Value`, `Sales_Units`) tienen una media alrededor de 33k / 422 unidades, pero se observan picos muy por encima de eso.
- Las principales palancas de medios (TV, YouTube, Facebook, Instagram) se miden en impresiones, mientras que las promociones se capturan con banderas y `Trade_Spend`.
- Hay controles operativos (precio, distribucion) y macroeconómicos (CPI, GDP, festivales, lluvia), lo cual permite separar baseline de incrementalidad.

Estas observaciones confirman que el archivo contiene tanto las variables de marketing necesarias como los controles que el MMM utiliza para distinguir causas.


## Cómo abordar un modelo MMM con este dataset
1. **Definir el KPI** (`Sales_Value` o `Sales_Units`) y decidir si el modelado se hace por SKU o por agregacion de marca/geografia.
2. **Transformaciones de marketing**: aplicar adstock a cada canal de impresiones y trade spend, luego una funcion de saturacion (Hill/log) para reflejar rendimientos decrecientes.
3. **Variables de control**: incluir precio (`Net_Price`, `MRP`), distribucion (`Weighted_Distribution`, `Numeric_Distribution`, `TDP`, `NOS`) y macrofactores (`CPI`, `GDP_Growth`, `Festival_Index`, `Rainfall_Index`). Agregar dummies de temporada/feriados si hay calendarios disponibles.
4. **Regularizacion y validacion**: usar Ridge o Elastic Net si hay alta correlacion entre canales; validar en bloques cronologicos para capturar estabilidad temporal.
5. **Interpretacion**: reportar baseline vs. lift, ROI por canal (incremental Sales_Value / gasto asociado), elasticidades log-log si se transforman los datos, e intervalos de confianza si se usa enfoque Bayesiano.

El archivo ya provee las palancas y controles tipicos, por lo que el siguiente paso es elegir la ventana temporal, normalizar las series y comenzar la optimizacion de los parametros adstock/saturacion.
